商品購入機能

sasakikeisuke-ft opened the pull request:
# What
ユーザーが商品を購入するとPAYJPを使用し決算が行われる。
商品を購入したユーザーが希望する郵送先の住所を保存する。

# why
商品を購入したユーザーの購入処理が正しく行われるため
郵送先住所を正しく保存することで、商品を購入したユーザーへ確実に郵送するため

GyazoGIF
*[正しく情報を記入すると、商品の購入ができる動画](https://gyazo.com/b81fa8bd335dad05026b022c661aa898)
*[入力に問題のある状態では商品の購入はできず、エラーメッセージが出力される動画](https://gyazo.com/8a7f2ccb50d642d42fe033140aade992)
*[ログアウト状態のユーザーが、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画](https://gyazo.com/9d2dcfcca9185ed5f69d77db49f8354f)
*[ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移する動画](https://gyazo.com/11375d8df54b1a03c442617c22ba1a71)
*[ログイン状態のユーザーが、URLを直接入力して売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画](https://gyazo.com/c4bcb9db85c3ee50eb9ce11ee9b36804)

*[売却済みの商品は、画像上に『sold out』の文字が表示される動画（商品一覧機能）](https://gyazo.com/94c3360dd448c1623fe4a881784635bb)
*[売却済みの商品は、画像上に『sold out』の文字が表示される動画（商品詳細機能）](https://gyazo.com/21384d7f76095f50ba880bbadcf03567)
*[ログイン状態の出品者でも、売却済みの商品に対しては「編集・削除ボタン」が表示されない動画](https://gyazo.com/d73ba331a2e1c0bdb08828383209dba0)
*[ログイン状態の出品者以外のユーザーでも、売却済みの商品に対しては「購入画面に進む」ボタンが表示されない動画](https://gyazo.com/4d3cd5c39d854aefe621d77dc5ebc893)
*[ログイン状態の出品者であっても、URLを直接入力して自身の売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画](https://gyazo.com/184a051cd4c30941c80b1629e77cf63b)

Gyazo
*[テスト結果の画像](https://gyazo.com/b6cc4d6d9d1bdbe26b6f2bddfa6390bc)

---

商品購入機能の初回コードレビュー依頼です。
一応は動くことを確認していますが、問題のある箇所もあるかと思います。
ご指摘をどうぞよろしくお願いします！

sasakikeisuke-ft committed:
ordersコントローラーの作成とアクションの仮置き

新しくordersコントローラーを作成しました。
またリンクが適切か検証できるようにindexアクションを仮置きしました。

sasakikeisuke-ft committed:
ordersコントローラーへのルーティングを設定

indexアクションに限定してordersコントローラーへのルーティングを設定しました。

sasakikeisuke-ft committed:
商品詳細ページから商品購入ページへのリンクを設定

sasakikeisuke-ft committed:
indexアクションに商品情報に関するインスタンス変数を設定

itemsテーブルからレコード情報を取得できるよう、indexアクションにインスタンス変数を設定しました。
ルーティングをネスト構造へ修正しました。
それに伴い商品詳細ページのリンクを修正しました。

sasakikeisuke-ft committed:
商品購入ページに商品情報が表示されるよう編集

商品購入ページに商品の画像、商品名、価格、支払いの負担などの商品情報が表示されるよう編集しました。

sasakikeisuke-ft committed:
orderモデルとaddressモデルを作成しました。

sasakikeisuke-ft committed:
orderモデルとaddressモデルのアソシエーションを記載

orderモデルとaddressモデルのアソシエーションを記載しました。
またそれに伴い、関係のあるuserモデル、itemモデル、ActiveHash:prefectureのアソシエーションも編集しました。

sasakikeisuke-ft committed:
マイグレーションファイルの編集

READMEに基づき、ordersテーブルとaddressesテーブルのマイグレーションファイルを編集しました。

sasakikeisuke-ft committed:
Formオブジェクトを作成

商品購入におけるレコード操作には、ordersテーブルとaddressesテーブルの二つのテーブルにまたがるため、Formオブジェクトを作成しました。

sasakikeisuke-ft committed:
商品購入ページの編集

商品購入ページのフォーム情報を編集しました。
それに伴い、ルーティングにcreateアクションを追加し、ordersコントローラーにcreateアクションを仮置きしました。

sasakikeisuke-ft committed:
createアクションの編集

ビューファイルからのフォーム情報を受け取れるよう、createアクションを編集しました。

sasakikeisuke-ft committed:
購入機能　途中

sasakikeisuke-ft committed:
Revert "購入機能　途中"

This reverts commit 083a58658714f60062b1e51903cba8aede0c22d5.

sasakikeisuke-ft committed:
Revert "Revert "購入機能　途中""

This reverts commit 6e50529e38f907fd0c95255068d12928248afd55.

sasakikeisuke-ft committed:
PAYJP導入

未導入だったPAYJPを導入しました。

sasakikeisuke-ft committed:
不必要なバリデーションの修正

空欄可能なbuildingカラムに空欄禁止のバリデーションがかかっていたため、修正しました。
また商品購入処理後に作成されるorder_idに対するバリデーションが記載されていたため削除しました。

sasakikeisuke-ft committed:
マイグレーションファイルの修正

addressesテーブルに関するマイグレーションファイルに誤りがあったため修正しました。
外部キーにより作成されるカラムがorderではなくuserになっていました。

sasakikeisuke-ft committed:
Formoオブジェクトの誤字を修正

order_addressのFormオブジェクトに誤字があったため修正しました。

sasakikeisuke-ft committed:
orderコントローラーのcreateアクションを作成

購入機能が正常に稼働するようcreateアクションを作成しました。
その際にストロングパラメーター取得メソッドと、PAYJP使用のためのメソッドを作成しました。

sasakikeisuke-ft committed:
共通コードの共有化

購入する商品に関するレコードを抽出するコードが重複していたため、プライベートに定義しbefore_actionで稼働するように編集しました。

sasakikeisuke-ft committed:
出品者本人ではないログインユーザーのみに限定

ログインユーザーのみ商品購入ページへ遷移できるよう調整しました。
また商品購入ページへ遷移できるのは出品者以外に制限しました。

sasakikeisuke-ft committed:
PAYJPのトークンに対しバリデーションを設定

sasakikeisuke-ft committed:
電話番号のバリデーション記載を変更

商品購入ページにおける電話番号の入力欄に10桁未満の数字を入力した場合、専用のエラーメッセージが出るようにバリデーションの記載を変更しました。

sasakikeisuke-ft committed:
売却済み商品の商品購入ページへ遷移した場合、トップページへ遷移するよう編集

対象商品が売却済みの状態で購入ページへ遷移しようとすると、トップページへ遷移するように調整しました。

sasakikeisuke-ft committed:
売却済みかどうかで商品詳細ページの表示が変更するように編集

商品詳細ページの画像が、売却済みならsold outが表示され、未売却であれば表示されないように編集しました。
また売却済みでは商品の編集、削除、購入ページへのリンクボタンが表示されないように調整しました。
実装に関する指示コメントは削除しました。

sasakikeisuke-ft committed:
売却済みかどうかでトップページの表示が変更されるように編集

トップページに並ぶ商品一覧に、売却済みであればsold outが表示され、未売却ならば表示されないように編集しました。
また実装内容の指示コメントは削除しました。

sasakikeisuke-ft committed:
バリデーションの脱字修正

バリデーションの脱字により電話番号が受け取れなくなっていたため修正しました。

sasakikeisuke-ft committed:
売却済み商品の編集、削除は行えないように編集。

出品者であっても商品が売却済みであった場合、商品情報の編集、削除が行えないように編集しました。

sasakikeisuke-ft committed:
Rubocop実行

Rubocopを実行し、リファクタリングを行いました。

sasakikeisuke-ft committed:
FactoryBot: order_address を作成

Formオブジェクトのorder_addressモデルのテストコードに必要なFactorybotを作成しました。

sasakikeisuke-ft committed:
order_addressモデルの単体テストコード作成

商品購入機能において情報が正しく保存されるか、order_addressモデルの単体テストコードを作成し確認しました。

sasakikeisuke-ft committed:
Rubocop実行

Rubocopを実行し、リファクタリングを行いました。

chobi555 commented on spec/factories/order_address.rb:
```diff
@@ -0,0 +1,14 @@
+FactoryBot.define do
+  factory :order_address do
+    item_id { Faker::Number.number(digits: 2) }
+    user_id { Faker::Number.number(digits: 2) }
```
外部キーを直接FactoryBotに記載するのは避ける方が無難です。 
今回購入において商品やユーザーを用意する必要があるのであれば、 
テストコードのbefore内でユーザーや商品のFactoryBotから生成しましょう。

chobi555 commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,97 @@
+        @order_address.valid?
+        expect(@order_address.errors.full_messages).to include("Phone number can't be blank")
+      end
+      it 'phone_numberに記号が含まれると登録できない' do
```
`電話番号は英数混合では登録できない`をテスト項目に追記しておきましょう。

chobi555 reviewed (commented):
レビュー依頼ありがとうございます！
コメント致しました。
確認後、修正をお願いいたします。

sasakikeisuke-ft committed:
修正1：電話番号のテストコードを追加

電話番号の保存に関するテストコードで「英字が含まれていると登録できない」を追加しました。

sasakikeisuke-ft committed:
修正2：外部キーを該当FactoryBotを使用し生成するように調整

外部キーをFactoryBotから直接生成している状態から、該当FactoryBotとbefore_actionを使用し、テストコード内で生成するように調整しました。
またcreateメソッドが重複していることによりmySQLのタイムアウトが生じることでエラーが発生しました。
これはsleepメソッドを使用することで解消しました。

sasakikeisuke-ft committed:
修正2追加：sleepメソッドの秒数調整

sleepの指定した秒数が短くテストコードの挙動が安定しなかったため、秒数を0.1秒に延長しました。

sasakikeisuke-ft commented:
What
商品購入機能　第一次修正

why
テストコードの内容をより正確にするため

GyazoGIF
動作に変更がないため割愛

Gyazo
*[テスト結果の画像](https://gyazo.com/ba53384566516122f38e114b052a6534)

---

ご指摘いただいた部分を修正しました。
再度確認をよろしくお願いします！

ken-1853 commented on app/models/order_address.rb:
```diff
@@ -0,0 +1,23 @@
+class OrderAddress
+  include ActiveModel::Model
+  attr_accessor :item_id, :user_id, :postal_code, :prefecture_id, :city, :address, :building, :phone_number, :order_id, :token
```
```
 attr_accessor :item_id, :user_id, :postal_code, :prefecture_id, :city, :address, :building, :phone_number, :order_id, :token
```
attr_accessor内にある

```ruby
:order_id
```

は不要のため、削除しましょう！
理由は、order_idはcreateメソッドの処理が正しく保存されてはじめて生成されるためです！

ken-1853 commented on spec/factories/order_address.rb:
```diff
@@ -0,0 +1,12 @@
+    address { Gimei.address.town.kanji }
+    building { '建物名' }
+    phone_number { Faker::PhoneNumber.subscriber_number(length: 11) }
+    order_id { Faker::Number.number(digits: 2) }
```
```
 order_id { Faker::Number.number(digits: 2) }
```
ファクトリーボット内においてorder_idは不要です。
理由としては、DBへ保存する処理をして初めてidが生成されるためです。

ken-1853 reviewed (commented):
レビュー提出ありがとうございます！
修正点を確認させていただきました。
コメントいたしましたので、ご確認お願いします！

今回テストコードの修正になりますので、最新のテストコードの実行結果をgyazo にてお送りいただいてもよろしいでしょうか
その後、再度提出フォームの送信をお願いします。

よろしくお願い致します！

sasakikeisuke-ft committed:
修正3：FactoryBotでorder_id生成を削除

order_idをFactoryBotで生成する必要はないため、記載を削除しました。

sasakikeisuke-ft committed:
修正4：attr_accessorからorder_idを削除しました。

order_idはDBに保存される際に自動で生成されるものでありattr_accessorへの記載は不要なため削除しました。

sasakikeisuke-ft commented:
What
商品購入機能　第二次修正

why
テストコードの内容をより正確にするため

GyazoGIF
動作に変更がないため割愛

Gyazo
*[テスト結果の画像](https://gyazo.com/e4e2cc5ba20658e77783087df476faa3)

---

ご指摘いただいた部分を修正しました。
添付した画像は新しく撮影したものになります。
再度確認をよろしくお願いします！

chobi555 commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,41 @@
+  end
+
+  def order_address_params
+    params.require(:order_address).permit(:postal_code, :prefecture_id, :city, :address, :building, :phone_number, :order_id).merge(
```
ストロングパラメーターにある`:order_id`を消しておきましょう。

chobi555 reviewed (commented):
レビュー依頼ありがとうございます！
一点細かい部分ではございますが、コメント致しました。
確認後、修正をお願いいたします。

sasakikeisuke-ft committed:
修正5：ストロングパラメーターよりorder_idを削除

DB保存時に生成されるorder_idはパラメーターとして送信されないため、ストロングパラメーターから削除しました。

sasakikeisuke-ft commented:
# What
商品購入機能　第三次修正

# why
コードの可読性向上のため

GyazoGIF
動作に変更がないため割愛

Gyazo
変更がないため割愛

---

ご指摘いただいた部分を修正しました。
テストコードばかりに着目し、コントローラーの修正は失念してしまいました。
お恥ずかしい限りです……。
再度確認をよろしくお願いします！

chobi555 commented:
レビュー依頼ありがとうございます！
LGTMとさせていただきます、ここまでの実装、本当にお疲れさまでした！

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
引き続き頑張っていきましょう！

sasakikeisuke-ft commented:
ありがとうございました！
残す実装もあと少し、頑張ってまいります！
どうぞよろしくお願いします！
